include/deal.II/matrix_free/operators.h

```diff
@@ -1089,29 +1112,64 @@ namespace MatrixFreeOperators
           const VectorizedArrayType *               in_array,
-          VectorizedArrayType *                     out_array) const
+          VectorizedArrayType *                     out_array,
+          const bool                                dyadic_coefficients) const
   {
-    const unsigned int given_degree =
-      fe_eval.get_shape_info().data[0].fe_degree;
     if (fe_degree > -1)
-      internal::CellwiseInverseMassMatrixImplFlexible<dim,
-                                                      VectorizedArrayType>::
-        template run<fe_degree>(
-          n_actual_components,
-          fe_eval.get_shape_info().data.front().inverse_shape_values_eo,
-          inverse_coefficients,
-          in_array,
-          out_array);
+      internal::CellwiseInverseMassMatrixImplFlexible<
+        dim,
+        VectorizedArrayType>::template run<fe_degree>(n_actual_components,
+                                                      fe_eval,
+                                                      inverse_coefficients,
+                                                      dyadic_coefficients,
+                                                      in_array,
+                                                      out_array);
     else
       internal::CellwiseInverseMassFactory<dim, VectorizedArrayType>::apply(
         n_actual_components,
-        given_degree,
-        fe_eval.get_shape_info().data.front().inverse_shape_values_eo,
+        fe_eval,
         inverse_coefficients,
+        dyadic_coefficients,
         in_array,
         out_array);
   }
 
+  template <int dim,
+            int fe_degree,
+            int n_components,
+            typename Number,
+            typename VectorizedArrayType>
+  inline void
+  CellwiseInverseMassMatrix<dim,
+                            fe_degree,
+                            n_components,
+                            Number,
+                            VectorizedArrayType>::
+    apply(const AlignedVector<Tensor<2, n_components, VectorizedArrayType>>
+            &                        inverse_dyadic_coefficients,
+          const VectorizedArrayType *in_array,
+          VectorizedArrayType *      out_array) const
+  {
+    const unsigned int dofs_per_component = inverse_dyadic_coefficients.size();
+    constexpr unsigned int n_tensor_components = n_components * n_components;
+
+    AlignedVector<VectorizedArrayType> inverse_coefficients(
+      dofs_per_component * n_tensor_components);
+
+    // Flatten the inverse dyadic coefficients into `inverse_coefficients`
+    {
+      auto begin = inverse_coefficients.begin();
+      for (unsigned int q = 0; q < dofs_per_component; ++q)
+        {
+          const auto end = std::next(begin, n_tensor_components);
+          inverse_dyadic_coefficients[q].unroll(begin, end);
+          begin = end;
+        }
+    }
+
+    apply(inverse_coefficients, n_components, in_array, out_array, true);
```

Comment:
Now you should be able to call
```suggestion
    apply(&inverse_dyadic_coefficients[0][0][0], n_components, in_array, out_array, true);
```
right? That would allow you to skip the rest of the code in this function.

Reply:
This is connected to the discussion https://github.com/dealii/dealii/pull/14850#issuecomment-1457901809m, right? My understanding is that the explicit unrolling is necessary, because elements of tensors aren't guaranteed to lie in contiguous memory. Do you think it is not necessary?

Reply:
Let us assume that tensors are contiguous in memory, copying would be too expensive. All of this happens inside internal interfaces of deal.II, so whenever there is a change in deal.II that would break the contiguous memory layout, we would need to address this here as well.

Reply:
Okay, one other problem is that the `apply()` method expects an `AlignedVector<VectorizedArrayType>`, so I just can't pass in the address. I couldn't find a way to construct an `AlignedVector` with a raw array.

Reply:
This is why I had suggested to switch this interface to either plain pointers or to `ArrayView`, in order to make sure we can pass in all options. Maybe `ArrayView` is good. All these are internal interfaces, so we can make the change without worrying about compatibility.

Reply:
Yep, I'm adopting `ArrayView` for the internal interfaces :+1: Thanks!